### PR TITLE
test: Initial aXe integration

### LINKED ISSUE #8641: tests: integrate aXe (accessibility engine)
aXe checks HTML + CSS for several accessibility-related issues and reports the status:
https://www.axe-core.org/

It can be ran in-browser, but also is able to be integrated into test suites.

- https://www.axe-core.org/examples/
- https://www.axe-core.org/examples/phantomjs.html

Accessibility checks should be run on every pull request.

We should also generate a list of errors across Cockpit master of issues to squash as well. These should probably be "grandparented in" for a few months as we fix them so we're not triggering errors in the meantime.

(However, if new code doesn't pass aXe, that should be considered an error and must be fixed before it is accepted.)

### PR DESCRIPTION
**Pull request by martinpitt**
Start running accessibility checks in our tests with
[aXe](https://www.axe-core.org). Add a `MachineCase.check_axe()` method
which generates an aXe report for the current frame and writes the JSON
reports into an attachment (names similar to screenshots and logs). This
gets heavily trimmed and gzipped to reduce the size from ~ 70 to 2 kB. 

Call this for the top-level navigation frame (in check-menu) and for the 
initial /system page (in check-system-info) for now. Once we get some
experience how to process, compare, and fix these, these calls will be
added to more tests.

To avoid injecting the axe check code into every test (which might skew
the results), introduce an `@enableAxe` decorator which does the 
injection. Every test that uses `check_axe()` must use that decorator.
Also, only do this on the default OS test for now (fedora-27 at the 
moment), as the results are going to be mostly identical between OSes
anyway.

Fixes #8641
------

~This is an initial PoC. I want to see if and how injecting the aXe helper affects the tests, and that attaching the two reports works as expected. However, the reports are currently much too large and verbose (roughly 40 kB each), and also rather hard to read. They need to be filtered to trim the information to what we need, similar to what the [CLI produces](https://i.imgur.com/HLc4x9u.png).~

 - [x] merge the two reports back into one
 - [x] trim reports
 - [x] add a decorator or similar to not impose the new inject on every test case
 - [x] fix attachments in CI

### CODE REVIEW

**Comment by martinpitt** [comment]
First test ran, but the report attachments are missing from the [results dir](https://fedorapeople.org/groups/cockpit/logs/pull-8780-20180306-162929-2747fec6-verify-fedora-27/), I added a TODO. There is no measurable runtime impact (this test ran for 3876s while the fedora-27 test in PR #8772 ran 4624s), but I still think we shouldn't do the inject into all tests to avoid possible corruptions. I added a TODO item for this too.

**Comment by martinpitt** [comment]
Attachment got fixed and reports merged, the [result dir](https://fedorapeople.org/groups/cockpit/logs/pull-8780-20180307-080404-34936aa0-verify-fedora-27/) now contains the [full report](https://fedorapeople.org/groups/cockpit/logs/pull-8780-20180307-080404-34936aa0-verify-fedora-27/Test-navigation-axe.json).

**Comment by martinpitt** [comment]
I added a decorator now  to limit the impact of this. @garret, can you please look into what information you need from the JSON?

**Comment by garrett** [comment]
I started looking into this last night with the JSON files, and I'm continuing today. I should have suggestions soon.

**Comment by garrett** [comment]
I was playing with what should be visible and what shouldn't.

Here's an attempt to filter the relevant information (formatted in a mess of HTML):
http://garrettlesage.com/cockpit-mockweb/axe/

The logic is basically:

- **violations** should be considered errors; these are clearly wrong and should be fixed
- **incomplete** should be considered warnings
  - these _can_ be just as wrong as violations, but require someone to look at the issue, as aXe couldn't determine what the problem is (and in some case, there might not be a problem)
  - at least, at a glance, these issues seem to be contrast and color related; this should be fixed, but a lot of the styles come from PatternFly — we could overwrite everything, but we probably should wait for a fix upstream (except in places where we ourselves control the color)
- **information to include**
  - short error message
  - link to details (offsite)
  - selectors (when there is more than one, the last looks the most useful for us — I'm not sure if it _happens_ to be the last here or is intentionally the last selector in the list — basically: the `#content > iframe` one seems to be reduntant and not useful; we should probably filter that out)
  - HTML snippet of the offending element
  - specific error message — however, this actually looks like it could be a bit redundant with the main error message
  - all the possible issues (where fixing the problem in one of the ways that best fits the way the element is used will fix the issue)
  - perhaps including the style information where it is applicable

This should hopefully prune the size down a good bit.

**Comment by martinpitt** [comment]
I retriggered a test after landing #8612.

Previous results: https://fedorapeople.org/groups/cockpit/logs/pull-8780-20180307-090808-423747e6-verify-fedora-27/
Current results: https://fedorapeople.org/groups/cockpit/logs/pull-8780-20180308-095353-423747e6-verify-fedora-27/

The "menu navigation" axe report indeed got a little smaller, but TestSystemInfo-testBasic-axe.json  *grew* by 7 kB.

**Comment by garrett** [comment]
I tried a few things:

- _several_ JSON diff comparing tools (they're all awful and fall apart when comparing this JSON)
- converting to YAML and comparing there with basic diff (this actually worked much better)
- ran the reformatted JSON files and also YAML through `wc -l` and determined that each shrunk a bit in number of lines

Results:

1. The changes in the JSON aren't as much as we have hoped for, but most of the accessibility fixes were spread across Cockpit. There was only 1 (IIRC) that affected the nav, so it's not too surprising.
2. Color contrast issues are something we do need to fix, but they're mainly at the PatternFly level. They're a large chunk of the warnings in the JSON.

**Comment by martinpitt** [comment]
@garrett: I did a first pass at trimming down the info, and this now appears quite readable: http://piware.de/tmp/axe/Test-navigation-axe.trimmed.json

Original report is 64 kB, trimmed report is 19 kB, gzipped is 2 kB. The latter seems fine for scaling, although it's less convenient to open (one needs to download and uncompress, firefox doesn't directly show .json.gz)

**Comment by garrett** [comment]
That... actually isn't bad at all.

It does remove some useful information, but it does provide enough to investigate the issue by grepping for code and/or navigating to the issue in browser-based development tools... and I guess that's what we need in these reports.

People can also still install the aXe browser add-on for Firefox or Chrome for more details to check things locally too.

**Comment by martinpitt** [comment]
I integrated the trimming as above now and added storing them in gzipped form. Now running tests on all OSes.

**Comment by martinpitt** [comment]
So this generally works, on e. g. https://fedorapeople.org/groups/cockpit/logs/pull-8780-20180320-152121-00658d9b-verify-ubuntu-1604/ we have the two aXe reports. However, I don't actually think it makes sense to run this on *every* test -- running it on just e. g. fedora-27 should suffice, right?

**Comment by martinpitt** [comment]
Pushed another change to only produce reports on the default OS (fedora-27 at the moment).

**Comment by martinpitt** [comment]
OK, now only [fedora-27](https://fedorapeople.org/groups/cockpit/logs/pull-8780-20180320-190404-0ebe9aab-verify-fedora-27/) has the aXe reports, none of the other tests.

File: test/common/testlib.py [hunk 1]
```diff
@@ -808,6 +810,61 @@ def check_browser_messages(self):
...
+        if suffix:
+            suffix = "-" + suffix
+        filename = "{0}{1}-axe.json.gz".format(label or self.label(), suffix)
+        with gzip.open(filename, "wb") as f:
+            f.write(json.dumps(report))
+        print("Wrote accessibility report to " + filename)
+        attach(filename)
+
+        # aXe triggers that *shrug*
+        self.allow_journal_messages("received invalid message without channel prefix")
```
Comment (larskarlitski) [inline test/common/testlib.py]: Oh, very suspicious. Do you have any idea how aXe could trigger sending an invalid message on the web socket?

File: test/common/testlib.py [hunk 2]
```diff
@@ -808,6 +810,61 @@ def check_browser_messages(self):
...
+        """Run aXe check on the currently active frame
+
+        The report gets written into an attachment
+        "<label>-axe-{violations,incomplete}.json". If you specify a suffix, it
+        will be appended to the file name, which is useful if you call this
+        more than once within one test.
+        """
+        # only run this on the default OS test, that's enough
+        if os.getenv("TEST_OS") not in [None, testvm.TEST_OS_DEFAULT]:
+            return
```
[the remark this answers is not in the document]
Comment (martinpitt) [reply test/common/testlib.py]: Note that right now we only collect reports at two places, and these do run everywhere.  Once this gets extended to e. g. the OSTree page, we of course need to run this on more OSes (or even all of them). But (1) until then we have hopefully fixed most of the "global" issues, thus making the reports much smaller; and (2) I wanted to start small, to not get overwhelmed by thousands of reports. This isn't meant to be the final solution anyway (e. g. we eventually want to gate on a11y regressions, not just report and forget them).

File: test/common/testlib.py (hunk 1 above)
Comment (martinpitt) [reply test/common/testlib.py]: No, sorry, I didn't look at that yet.

Comment (larskarlitski) [reply test/common/testlib.py]: I'm fine merging it with this, but I'm afraid we'll forget about looking at it. Shall I open an issue about it?

**Pull request merged**